File: stockyapi/Services/PortfolioService.cs
using System.Security.Claims;
using stockyapi.Repository.Portfolio;
using stockyapi.Requests;
using stockyapi.Responses;
using stockymodels.models;

namespace stockyapi.Services;

public interface IPortfolioService
{
    public Task<UserPortfolioResponse> GetUserPortfolio();
    public Task<BuyTickerResponse> BuyTickerInPortfolio(BuyTickerRequest request);
    public Task<SellTickerResponse> SellTickerInPortfolio(SellTickerRequest request);
    public Task<AddFundsResponse> AddFunds(AddFundsRequest request);
    public Task<SetFundsResponse> SetFunds(SetFundsRequest request);
    public Task<SubtractFundsResponse> SubtractFunds(SubtractFundsRequest request);
}

public class PortfolioService : IPortfolioService
{
    private readonly IHttpContextAccessor _httpContextAccessor;
    private readonly IPortfolioRepository _portfolioRepository;

    public PortfolioService(
        IHttpContextAccessor httpContextAccessor,
        IPortfolioRepository portfolioRepository)
    {
        _httpContextAccessor = httpContextAccessor;
        _portfolioRepository = portfolioRepository;
    }

    public async Task<UserPortfolioResponse> GetUserPortfolio()
    {
        var userId = _httpContextAccessor.HttpContext?.User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (userId == null)
        {
            return new UserPortfolioResponse
            {
                Success = false,
                StatusCode = 401,
                Message = "User not authenticated"
            };
        }

        var portfolio = await _portfolioRepository.GetByUserIdAsync(Guid.Parse(userId));
        if (portfolio == null)
        {
            return new UserPortfolioResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "User portfolio not found"
            };
        }

        var portfolioItems = portfolio.StockHoldings.Select(holding => new PortfolioItem
        {
            Symbol = holding.Symbol,
            Quantity = holding.Shares,
            AverageBuyPrice = holding.AverageCost,
            CurrentPrice = holding.CurrentPrice,
            TotalValue = holding.Shares * holding.CurrentPrice,
            ProfitLoss = (holding.CurrentPrice - holding.AverageCost) * holding.Shares,
            ProfitLossPercentage = holding.AverageCost != 0
                ? ((holding.CurrentPrice - holding.AverageCost) / holding.AverageCost) * 100
                : 0,
            LastUpdatedTime = holding.UpdatedAt.ToString("yyyy-MM-dd HH:mm:ss") ?? holding.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss")
        }).ToList();

        return new UserPortfolioResponse
        {
            Success = true,
            StatusCode = 200,
            Data = new PortfolioData
            {
                TotalValue = portfolio.TotalValue,
                CashBalance = portfolio.CashBalance,
                InvestedAmount = portfolio.InvestedAmount,
                Items = portfolioItems
            }
        };
    }

    public async Task<BuyTickerResponse> BuyTickerInPortfolio(BuyTickerRequest request)
    {
        var userId = _httpContextAccessor.HttpContext?.User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (userId == null)
        {
            return new BuyTickerResponse
            {
                Success = false,
                StatusCode = 401,
                Message = "User not authenticated"
            };
        }

        var portfolio = await _portfolioRepository.GetByUserIdAsync(Guid.Parse(userId));
        if (portfolio == null)
        {
            return new BuyTickerResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Portfolio not found"
            };
        }

        if (request.Quantity <= 0)
        {
            return new BuyTickerResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Quantity must be greater than 0"
            };
        }

        if (request.Price <= 0)
        {
            return new BuyTickerResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Price must be greater than 0"
            };
        }

        // Validate purchase
        var totalCost = request.Price * request.Quantity;
        if (totalCost > portfolio.CashBalance)
        {
            return new BuyTickerResponse
            {
                Success = false,
                StatusCode = 400,
                Message = "Insufficient funds"
            };
        }

        // Create transaction
        var transaction = new TransactionModel
        {
            PortfolioId = portfolio.Id,
            Symbol = request.Symbol,
            Type = TransactionType.Buy,
            Shares = request.Quantity,
            Price = request.Price,
            TotalAmount = totalCost,
            Status = TransactionStatus.Completed,
            OrderType = OrderType.Market,
            LastPriceUpdate = DateTime.UtcNow
        };

        // Update portfolio balance
        portfolio.CashBalance -= totalCost;
        portfolio.InvestedAmount += totalCost;

        // Update or create stock holding
        var existingHolding = await _portfolioRepository.GetHoldingAsync(portfolio.Id, request.Symbol);
        if (existingHolding != null)
        {
            // Update existing holding
            var newTotalShares = existingHolding.Shares + request.Quantity;
            var newTotalCost = (existingHolding.AverageCost * existingHolding.Shares) + totalCost;
            existingHolding.Shares = newTotalShares;
            existingHolding.AverageCost = newTotalCost / newTotalShares;
            existingHolding.CurrentPrice = request.Price;
            await _portfolioRepository.UpdateHoldingAsync(existingHolding);
        }
        else
        {
            // Create new holding
            var newHolding = new StockHoldingModel
            {
                PortfolioId = portfolio.Id,
                Symbol = request.Symbol,
                Shares = request.Quantity,
                AverageCost = request.Price,
                CurrentPrice = request.Price
            };
            await _portfolioRepository.AddHoldingAsync(newHolding);
        }

        // Save transaction and update portfolio
        await _portfolioRepository.AddTransactionAsync(transaction);
        await _portfolioRepository.UpdateAsync(portfolio);

        return new BuyTickerResponse
        {
            Success = true,
            StatusCode = 200,
            Data = new BuyTickerData
            {
                Symbol = request.Symbol,
                Quantity = request.Quantity,
                Price = request.Price,
                TotalCost = totalCost,
                RemainingCashBalance = portfolio.CashBalance,
                TransactionTime = DateTime.UtcNow,
                TransactionId = transaction.Id.ToString(),
                Status = transaction.Status
            }
        };
    }

    public async Task<SellTickerResponse> SellTickerInPortfolio(SellTickerRequest request)
    {
        var userId = _httpContextAccessor.HttpContext?.User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (userId == null)
        {
            return new SellTickerResponse
            {
                Success = false,
                StatusCode = 401,
                Message = "User not authenticated"
            };
        }

        var portfolio = await _portfolioRepository.GetByUserIdAsync(Guid.Parse(userId));
        if (portfolio == null)
        {
            return new SellTickerResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Portfolio not found"
            };
        }

        var total = request.Price * request.Quantity;

        // Create transaction
        var transaction = new TransactionModel
        {
            PortfolioId = portfolio.Id,
            Symbol = request.Symbol,
            Type = TransactionType.Sell,
            Shares = request.Quantity,
            Price = request.Price,
            TotalAmount = total,
            Status = TransactionStatus.Completed,
            OrderType = OrderType.Market,
            LastPriceUpdate = DateTime.UtcNow
        };

        // Update portfolio
        portfolio.CashBalance += total;
        portfolio.InvestedAmount -= total;

        // Update stock holding
        var existingHolding = await _portfolioRepository.GetHoldingAsync(portfolio.Id, request.Symbol);
        if (existingHolding == null)
        {
            return new SellTickerResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Stock is not owned"
            };
        }

        if (existingHolding.Shares < request.Quantity)
        {
            return new SellTickerResponse
            {
                Success = false,
                StatusCode = 400,
                Message = "Insufficient shares to sell"
            };
        }

        // Update holding
        var newTotalShares = existingHolding.Shares - request.Quantity;
        var newTotalCost = (existingHolding.AverageCost * existingHolding.Shares) - total;
        existingHolding.Shares = newTotalShares;
        existingHolding.AverageCost = newTotalShares == 0 ? 0 : newTotalCost / newTotalShares;
        existingHolding.CurrentPrice = request.Price;

        // Save changes
        await _portfolioRepository.AddTransactionAsync(transaction);
        await _portfolioRepository.UpdateHoldingAsync(existingHolding);
        await _portfolioRepository.UpdateAsync(portfolio);

        return new SellTickerResponse
        {
            Success = true,
            StatusCode = 200,
            Data = new SellTickerData
            {
                Symbol = request.Symbol,
                Quantity = request.Quantity,
                Price = request.Price,
                TotalCost = total,
                RemainingCashBalance = portfolio.CashBalance,
                TransactionTime = DateTime.UtcNow,
                TransactionId = transaction.Id.ToString(),
                Status = transaction.Status
            }
        };
    }

    public async Task<AddFundsResponse> AddFunds(AddFundsRequest request)
    {
        var userId = _httpContextAccessor.HttpContext?.User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (userId == null)
        {
            return new AddFundsResponse
            {
                Success = false,
                StatusCode = 401,
                Message = "User not authenticated"
            };
        }

        var portfolio = await _portfolioRepository.GetByUserIdAsync(Guid.Parse(userId));
        if (portfolio == null)
        {
            return new AddFundsResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Portfolio not found"
            };
        }

        var newBalance = await _portfolioRepository.AddCashAsync(portfolio.Id, request.Amount);
        return new AddFundsResponse
        {
            Success = true,
            StatusCode = 200,
            Data = new AddFundsData
            {
                NewBalance = newBalance
            }
        };
    }

    public async Task<SubtractFundsResponse> SubtractFunds(SubtractFundsRequest request)
    {
        var userId = _httpContextAccessor.HttpContext?.User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (userId == null)
        {
            return new SubtractFundsResponse
            {
                Success = false,
                StatusCode = 401,
                Message = "User not authenticated"
            };
        }

        var portfolio = await _portfolioRepository.GetByUserIdAsync(Guid.Parse(userId));
        if (portfolio == null)
        {
            return new SubtractFundsResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Portfolio not found"
            };
        }

        var newBalance = await _portfolioRepository.SubtractCashAsync(portfolio.Id, request.Amount);
        return new SubtractFundsResponse
        {
            Success = true,
            StatusCode = 200,
            Data = new SubtractFundsData
            {
                NewBalance = newBalance
            }
        };
    }

    public async Task<SetFundsResponse> SetFunds(SetFundsRequest request)
    {
        var userId = _httpContextAccessor.HttpContext?.User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (userId == null)
        {
            return new SetFundsResponse
            {
                Success = false,
                StatusCode = 401,
                Message = "User not authenticated"
            };
        }

        var portfolio = await _portfolioRepository.GetByUserIdAsync(Guid.Parse(userId));
        if (portfolio == null)
        {
            return new SetFundsResponse
            {
                Success = false,
                StatusCode = 404,
                Message = "Portfolio not found"
            };
        }

        var newBalance = await _portfolioRepository.SetCashAsync(portfolio.Id, request.Amount);
        return new SetFundsResponse
        {
            Success = true,
            StatusCode = 200,
            Data = new SetFundsData
            {
                NewBalance = newBalance
            }
        };
    }
}

File: stockyui/src/services/portfolio.service.ts
import { BaseService } from './base.service';
import { StockyApi } from './generated/stockyapi';

export class PortfolioService extends BaseService {
  // API Methods
  async getPortfolio() {
    try {
      return await this.api.portfolio();
    } catch (error) {
      console.error('Error fetching portfolio:', error);
      return {
        success: false,
        statusCode: 500,
        message: 'Failed to fetch portfolio'
      } as StockyApi.UserPortfolioResponse;
    }
  }

  async buyTicker(request: StockyApi.BuyTickerRequest) {
    try {
      return await this.api.buy(request);
    } catch (error) {
      console.error('Error buying ticker:', error);
      return {
        success: false,
        statusCode: 500,
        message: 'Failed to buy ticker'
      } as StockyApi.BuyTickerResponse;
    }
  }

  async sellTicker(request: StockyApi.SellTickerRequest) {
    try {
      return await this.api.sell(request);
    } catch (error) {
      console.error('Error selling ticker:', error);
      return {
        success: false,
        statusCode: 500,
        message: 'Failed to sell ticker'
      } as StockyApi.SellTickerResponse;
    }
  }

  async addFunds(request: StockyApi.AddFundsRequest) {
    try {
      return await this.api.addfunds(request);
    } catch (error) {
      console.error('Error adding funds:', error);
      return {
        success: false,
        statusCode: 500,
        message: 'Failed to add funds'
      } as StockyApi.AddFundsResponse;
    }
  }

  async subtractFunds(request: StockyApi.SubtractFundsRequest) {
    try {
      return await this.api.subtractfunds(request);
    } catch (error) {
      console.error('Error subtracting funds:', error);
      return {
        success: false,
        statusCode: 500,
        message: 'Failed to subtract funds'
      } as StockyApi.SubtractFundsResponse;
    }
  }

  async setFunds(request: StockyApi.SetFundsRequest) {
    try {
      return await this.api.setfunds(request);
    } catch (error) {
      console.error('Error setting funds:', error);
      return {
        success: false,
        statusCode: 500,
        message: 'Failed to set funds'
      } as StockyApi.SetFundsResponse;
    }
  }

  async deleteTicker(request: StockyApi.DeleteTickerRequest) {
    try {
      return await this.api.delete(request);
    } catch (error) {
      console.error('Error deleting ticker:', error);
      return {
        success: false,
        statusCode: 500,
        message: 'Failed to delete ticker'
      } as StockyApi.DeleteTickerResponse;
    }
  }

  // Portfolio Analysis
  calculateTotalValue(portfolio: StockyApi.PortfolioData): number {
    return portfolio.totalValue || 0;
  }

  calculateProfitLoss(portfolio: StockyApi.PortfolioData): number {
    return portfolio.items?.reduce((total, item) => {
      return total + (item.profitLoss || 0);
    }, 0) || 0;
  }

  getTopPerformers(portfolio: StockyApi.PortfolioData, limit: number = 5): StockyApi.PortfolioItem[] {
    return [...(portfolio.items || [])]
      .sort((a, b) => (b.profitLossPercentage || 0) - (a.profitLossPercentage || 0))
      .slice(0, limit);
  }

  getWorstPerformers(portfolio: StockyApi.PortfolioData, limit: number = 5): StockyApi.PortfolioItem[] {
    return [...(portfolio.items || [])]
      .sort((a, b) => (a.profitLossPercentage || 0) - (b.profitLossPercentage || 0))
      .slice(0, limit);
  }

  calculateAllocation(portfolio: StockyApi.PortfolioData): Map<string, number> {
    const totalValue = this.calculateTotalValue(portfolio);
    const allocation = new Map<string, number>();

    portfolio.items?.forEach(item => {
      if (item.symbol && item.totalValue) {
        const percentage = (item.totalValue / totalValue) * 100;
        allocation.set(item.symbol, percentage);
      }
    });

    return allocation;
  }
}